dcos/http.py

```diff
@@ -311,55 +314,87 @@ def _get_basic_auth_credentials(username, hostname):
-                           None)
-        if basic_realm:
-            realm = basic_realm.split('=')[-1].strip(' \'\"').lower()
-            return realm
+        scheme = next((auth_type.rstrip().lower() for auth_type in auths
+                       if auth_type.rstrip().lower().startswith("basic") or
+                       auth_type.rstrip().lower().startswith("acsjwt")),
+                      None)
+        if scheme:
+            scheme_info = scheme.split("=")
+            auth_scheme = scheme_info[0].split(" ")[0].lower()
+            realm = scheme_info[-1].strip(' \'\"').lower()
+            return auth_scheme, realm
         else:
             return None
     else:
         return None
 
 
-def _get_http_auth_credentials(response):
-    """Get authentication credentials required by server
+def _get_http_auth(response, url, auth_scheme):
+    """Get authentication mechanism required by server
 
     :param response: requests.response
     :type response: requests.Response
-    :returns: HTTPBasicAuth
-    :rtype: HTTPBasicAuth
+    :param url: parsed request url
+    :type url: str
+    :param auth_scheme: str
+    :type auth_scheme: str
+    :returns: AuthBase
+    :rtype: AuthBase
     """
 
-    parsed_url = urlparse(response.url)
-    hostname = parsed_url.hostname
-    user = parsed_url.username
+    hostname = url.hostname
+    username = url.username
 
     if 'www-authenticate' in response.headers:
-        realm = _get_realm(response)
-        if realm:
-            return _get_basic_auth_credentials(user, hostname)
-        else:
+        if auth_scheme not in ['basic', 'acsjwt']:
             msg = ("Server responded with an HTTP 'www-authenticate' field of "
                    "'{}', DCOS only supports 'Basic'".format(
                        response.headers['www-authenticate']))
             raise DCOSException(msg)
+
+        username, password = _get_auth_credentials(username, hostname)
+        if auth_scheme == 'basic':
+            return HTTPBasicAuth(username, password)
+        else:
+            return DCOSAcsAuth(username, password)
     else:
         msg = ("Invalid HTTP response: server returned an HTTP 401 response "
                "with no 'www-authenticate' field")
         raise DCOSException(msg)
+
+
+class DCOSAcsAuth(AuthBase):
+    """Invokes DCOS Authentication flow for given Request object."""
+    def __init__(self, uid, password):
+        self.uid = uid
+        self.password = password
+
+    def __call__(self, r):
+        dcos_url = util.get_config_vals(
+            ['core.dcos_url'], util.get_config())[0]
+        url = urllib.parse.urljoin(dcos_url, 'acs/api/v1/auth/login')
+        creds = {"uid": self.uid, "password": self.password}
+
+        # using private method here, so we don't retry on this request
+        # error here will be bubbled up to _request_with_auth
+        response = _request('post', url, json=creds)
+        if response.status_code == 200:
```

Comment:
I think I get it. If the header is not set then the following call will get fail with a `400`.